once/__init__.py

```diff
@@ -46,78 +51,145 @@ def _wrapped_function_type(func: collections.abc.Callable) -> _WrappedFunctionTy
+    # definition time, so we force the caller to pass it in. But, if we're in debug mode, why not
+    # check it again?
+    assert fn_type == _wrapped_function_type(func)
+    wrapped: collections.abc.Callable
+    if fn_type == _WrappedFunctionType.ASYNC_GENERATOR:
+
+        async def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            async with once_base.async_lock:
+                if not once_base.called:
+                    once_base.return_value = _iterator_wrappers.AsyncGeneratorWrapper(
+                        func, *args, **kwargs
+                    )
+                    once_base.called = True
+                return_value = once_base.return_value
+            next_value = None
+            iterator = return_value.yield_results()
+            while True:
+                try:
+                    next_value = yield await iterator.asend(next_value)
+                except StopAsyncIteration:
+                    return
+
+    elif fn_type == _WrappedFunctionType.ASYNC_FUNCTION:
+
+        async def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            async with once_base.async_lock:
+                if not once_base.called:
+                    once_base.return_value = await func(*args, **kwargs)
+                    once_base.called = True
+                return once_base.return_value
+
+    elif fn_type == _WrappedFunctionType.SYNC_FUNCTION:
+
+        def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            with once_base.lock:
+                if not once_base.called:
+                    once_base.return_value = func(*args, **kwargs)
+                    once_base.called = True
+                return once_base.return_value
+
+    elif fn_type == _WrappedFunctionType.SYNC_GENERATOR:
+
+        def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            with once_base.lock:
+                if not once_base.called:
+                    once_base.return_value = _iterator_wrappers.GeneratorWrapper(
+                        func, *args, **kwargs
+                    )
+                    once_base.called = True
+                iterator = once_base.return_value
+            yield from iterator.yield_results()
+
+    else:
+        raise NotImplementedError()
+
+    functools.update_wrapper(wrapped, func)
+    return wrapped
+
+
+def _once_factory(is_async: bool, per_thread: bool) -> _ONCE_FACTORY_TYPE:
+    if not per_thread:
+        singleton_once = _OnceBase(is_async)
+        return lambda: singleton_once
+
+    per_thread_onces = threading.local()
+
+    def _get_once_per_thread():
+        # Read then modify is thread-safe without a lock because `threading.local` is only present
+        # in the currently active thread, which cannot race with itself!
```

Comment:
Just a nit: it's not so much that `threading.local` is only present in the currently active thread so much as that each thread sees its own copy of `per_thread_onces`. IMO a more clear and correct way to phrase it might be:
```python
# Read then modify is thread-safe without a lock because each thread sees its own
# copy of `per_thread_onces` thanks to `threading.local`, and each thread cannot race with
# itself!
```

Reply:
done